buildstockbatch/cloud/docker_base.py

```diff
@@ -56,30 +71,171 @@ def docker_image(self):
-            unique_epws: A dictionary mapping from the hash of weather files to a list of filenames
-                with that hashed value. Only the first in each list is written to tmppath, so
-                this can be used to recreate the other files later.
+        :param batch_info: A :class:`DockerBatchBase.BatchInfo` containing information about the job
         """
-        # Generate buildstock.csv
-        buildstock_csv_filename = self.sampler.run_sampling()
+        raise NotImplementedError
+
+    def run_batch(self):
+        """Prepare and start a Batch job on the cloud to run simulations.
+
+        This does all the cloud-agnostic prep (such as preparing weather files, assets, and job
+        definition), delegating to the implementations to upload those files to the cloud (using
+        (:func:`upload_batch_files_to_cloud` and :func:`copy_files_at_cloud`), and then calls the
+        implementation's :func:`start_batch_job` to actually create and start the batch job.
+        """
+        with tempfile.TemporaryDirectory(prefix="bsb_") as tmpdir:
+            tmppath = pathlib.Path(tmpdir)
+            epws_to_copy, batch_info = self._run_batch_prep(tmppath)
+
+            # Copy all the files to cloud storage
+            logger.info("Uploading files for batch...")
+            self.upload_batch_files_to_cloud(tmppath)
+
+            logger.info("Copying duplicate weather files...")
+            self.copy_files_at_cloud(epws_to_copy)
+
+        self.start_batch_job(batch_info)
+
+    def _run_batch_prep(self, tmppath):
+        """Do preparation for running the Batch jobs on the cloud, including producing and uploading
+        files to the cloud that the batch jobs will use.
+
+        This includes:
+            - Weather files (:func:`_prep_weather_files_for_batch`)
+            - Assets (:func:`_prep_assets_for_batch`)
+            - Sampling, and splitting the samples into (at most) ``self.batch_array_size`` batches
+              (:func:`_prep_jobs_for_batch`)
+
+        Those functions place their files to be uploaded into ``tmppath``, and then this will upload
+        them to the cloud using (:func:`upload_batch_files_to_cloud`).
+
+        Duplicate weather files will have been excluded from ``tmppath``, and this will use
+        (:func:`copy_files_at_cloud`) to copy those files from-cloud-to-cloud (instead of uploading
+        them).
+
+        ``self.weather_dir`` must exist before calling this method. This is where weather files are
+        stored temporarily.
+
+        This takes ``tmppath`` (rather than managing itself) for testability (so test can manage and
+        inspect the contents of the tmppath).
+
+        :returns: DockerBatchBase.BatchInfo
+        """
+
+        # Weather files
+        logger.info("Prepping weather files...")
+        epws_to_copy = self._prep_weather_files_for_batch(tmppath)
+
+        # Assets
+        self._prep_assets_for_batch(tmppath)
```

Comment:
This needs to run after `_prep_jobs_for_batch`, because the assets being uploaded. include `buildstock.csv` that is produced by `run_sampling()`.

Reply:
Good point, @nweires.

Reply:
Cherry-picked the fix from rewiringamerica/buildstockbatch#41